### PR TITLE
ARTEMIS-5142 setting 0 for min/max expiry-delivery not working as expected

### PR DESCRIPTION
**Pull request by jbertram**

### CODE REVIEW

File: artemis-jdbc-store/src/main/java/org/apache/activemq/artemis/jdbc/store/file/JDBCSequentialFile.java [hunk 1]
```diff
@@ -92,6 +92,7 @@ public class JDBCSequentialFile implements SequentialFile {
       this.filename = filename;
       this.extension = filename.contains(".") ? filename.substring(filename.lastIndexOf(".") + 1, filename.length()) : "";
       this.executor = executor;
+      new Exception(driver.toString()).printStackTrace();
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-jdbc-store/src/main/java/org/apache/activemq/artemis/jdbc/store/file/JDBCSequentialFile.java]: Indeed!

File: pom.xml
Comment (gemmellr) [inline pom.xml:1]: These changes seem unrelated

**Comment by gemmellr** [comment]
The min and max settings aren't listed in the address-settings doc, might be good to add them with this.

(They are covered in the expiry section referenced from the address setting doc for the other expiry setting).

File: pom.xml
Comment (jbertram) [reply pom.xml:1]: Fixed.

**Comment by jbertram** [comment]
@gemmellr, I took care of the documentation.

File: artemis-server/src/test/java/org/apache/activemq/artemis/core/postoffice/impl/ExpiryTest.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.core.postoffice.impl;
+
+import org.apache.activemq.artemis.api.core.Message;
+import org.apache.activemq.artemis.core.settings.impl.AddressSettings;
+import org.junit.Test;
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/test/java/org/apache/activemq/artemis/core/postoffice/impl/ExpiryTest.java]: Fixed.

File: artemis-server/src/test/java/org/apache/activemq/artemis/core/postoffice/impl/ExpiryTest.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import org.mockito.Mockito;
+
+public class ExpiryTest {
+
+   @Test
+   public void testZeroMaxExpiryDelayWhenExpirationSet() {
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/test/java/org/apache/activemq/artemis/core/postoffice/impl/ExpiryTest.java]: I added the ones you mentioned plus a few more.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java [hunk 4]
```diff
@@ -1366,17 +1367,32 @@ private static void applyExpiryDelay(Message message, AddressSettings settings)
...
-            message.setExpiration(System.currentTimeMillis() + maxExpiration);
+         // if the incoming message has NO expiration then apply the max if set and if not set then apply the min if set
+         if (message.getExpiration() == 0) {
+            if (maxExpiration != AddressSettings.DEFAULT_MAX_EXPIRY_DELAY && maxExpiration != 0) {
+               message.setExpiration(getExpirationToSet(maxExpiration));
+            } else if (minExpiration != AddressSettings.DEFAULT_MIN_EXPIRY_DELAY && minExpiration != 0) {
+               message.setExpiration(getExpirationToSet(minExpiration));
+            }
+         } else if (maxExpiration != AddressSettings.DEFAULT_MAX_EXPIRY_DELAY && message.getExpiration() > (System.currentTimeMillis() + maxExpiration)) {
+            message.setExpiration(getExpirationToSet(maxExpiration));
```
Comment (gemmellr) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java:1378]: So this and the segment below seems to try handling the 'max/min expiration delay 0' special case, setting the expiration to 0. One covers \>current and the other covers \<current.

However if setting both settings to 0, presumably for an implied do-no-expiry-ever, then this setup will currently miss the case where the value exactly matches the current time reference, which will then be left unchanged and be allowed to expire which could be considered against the expectations for this special case?

**Comment by gemmellr** [comment]
In addition to the above comments on potential for mismatch between doc and/or expectations and the actual behaviour...I pushed another commit with some more tests, and tweaks to some of the existing tests, to more strictly verify the resulting expiration value set is in the expected range. See what you think.

**Comment by gemmellr** [comment]
Other than those bits I commented on where its clearly not ok? :)

**Comment by clebertsuconic** [comment]
@gemmellr I said.. loTm... not clearly to you :)

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java [hunk 5]
```diff
@@ -1366,17 +1367,32 @@ private static void applyExpiryDelay(Message message, AddressSettings settings)
...
-
-         if (maxExpiration != AddressSettings.DEFAULT_MAX_EXPIRY_DELAY && (message.getExpiration() == 0 || message.getExpiration() > (System.currentTimeMillis() + maxExpiration))) {
-            message.setExpiration(System.currentTimeMillis() + maxExpiration);
+         // if the incoming message has NO expiration then apply the max if set and if not set then apply the min if set
+         if (message.getExpiration() == 0) {
+            if (maxExpiration != AddressSettings.DEFAULT_MAX_EXPIRY_DELAY && maxExpiration != 0) {
+               message.setExpiration(getExpirationToSet(maxExpiration));
+            } else if (minExpiration != AddressSettings.DEFAULT_MIN_EXPIRY_DELAY && minExpiration != 0) {
+               message.setExpiration(getExpirationToSet(minExpiration));
+            }
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java:1376]: The idea with `&& maxExpiration != 0` was simply to avoid setting the expiration to `0` when it already was `0`. However, I see the issue with the logic that you mention so I'm taking it out. Good catch!

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java (hunk 4 above)
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java:1378]: Fixing!

**Comment by jbertram** [comment]
@gemmellr, the extra/modified tests in your commit look good. How do you want to merge them? Do you want to rebase on the `push -f` I'm about to do or do you want me to just incorporate them into my own?

**Comment by gemmellr** [comment]
Feel free to squash them

**Comment by jbertram** [comment]
I decided to keep your commit. Merge when ready.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (gemmellr) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java]: I didnt think it should be removed, just otherwise fixed.

Allowing it to call setExpirartion(0) registers the message as having been changed, when it has not been. I think you were correct to avoid that case occurring.

[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java]: Done.

File: artemis-server/src/test/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImplTest.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+public class PostOfficeImplTest {
+
+   private static final int EXPIRATION_DELTA = 5000;
+
+   @Test
+   public void testZeroMaxExpiryDelayWhenExpirationNotSet() {
+      Message mockMessage = Mockito.mock(Message.class);
+      Mockito.when(mockMessage.getExpiration()).thenReturn(0L);
+      PostOfficeImpl.applyExpiryDelay(mockMessage, new AddressSettings().setMaxExpiryDelay(0L));
+      Mockito.verify(mockMessage).setExpiration(Mockito.anyLong());
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/test/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImplTest.java]: Fixed.

**Comment by gemmellr** [comment]
Also, I noticed the regular expiry-delay (not min/max) setting has the same '0 makes it set expire now' behaviour because it will set the current time as the expiration. It is subtly different from the min/max case in that the setting only ever applies to messages with an existing expiration, so setting expiry-delay to 0 really makes no sense to begin with, so it could more reasonably just be left as it is, but I wonder if it should also special-case it like min/max now do?

**Comment by jbertram** [comment]
> ...setting expiry-delay to 0 really makes no sense to begin with...

That surely means someone will do it. I'll add a test and update the logic to deal with this.

File: artemis-server/src/test/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImplTest.java [hunk 7]
```diff
@@ -99,6 +99,14 @@ public void testZeroMinExpiryDelayWhenExpirationSet() {
       Mockito.verify(mockMessage).setExpiration(0L);
    }
 
+   @Test
+   public void testZeroMinExpiryDelayWhenExpirationSetInFuture() {
+      Message mockMessage = Mockito.mock(Message.class);
+      Mockito.when(mockMessage.getExpiration()).thenReturn(System.currentTimeMillis() + 500_000L);
+      PostOfficeImpl.applyExpiryDelay(mockMessage, new AddressSettings().setMinExpiryDelay(0L));
+      Mockito.verify(mockMessage).setExpiration(0L);
+   }
```
Comment (jbertram) [inline artemis-server/src/test/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImplTest.java:108]: Thinking about this more...I'm not sure a `min-expiry-delay` of `0` even makes sense. The overall goal of this change is to enable users to never expire incoming messages regardless of the message's existing expiration. If they want to do that then they should set `max-expiry-delay` to `0` and leave `expiry-delay` and `min-expiry-delay` unset (i.e. `-1`). 

Setting `min-expiry-delay` to `0` is really just meaningless. A message will not realistically have an expiration below `0` and if there's no max set (as in this test) then anything above `0` should remain unchanged.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java [hunk 8]
```diff
@@ -1366,17 +1367,36 @@ private static void applyExpiryDelay(Message message, AddressSettings settings)
...
+               }
+            } else if (minExpiration != AddressSettings.DEFAULT_MIN_EXPIRY_DELAY) {
+               if (minExpiration != 0) {
+                  message.setExpiration(getExpirationToSet(minExpiration));
+               }
+            }
+         } else if (maxExpiration != AddressSettings.DEFAULT_MAX_EXPIRY_DELAY && message.getExpiration() >= (System.currentTimeMillis() + maxExpiration)) {
+            message.setExpiration(getExpirationToSet(maxExpiration));
+         } else if (minExpiration != AddressSettings.DEFAULT_MIN_EXPIRY_DELAY && (minExpiration == 0 || message.getExpiration() < (System.currentTimeMillis() + minExpiration))) {
+            message.setExpiration(getExpirationToSet(minExpiration));
```
[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java:1384]: Hmmm...So maybe the current behavior of `min-expiry-delay` and `max-expiry-delay` is actually correct in a strict semantic interpretation and instead of making `0` behave in a special way we should just add a new address-setting to strip the expiration from incoming messages so they never expire categorically which is the semantic that the Jira is really aimed at.

[the remark this answers is not in the document]
Comment (jbertram) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java:1384]: I put together #5390 which does this very thing. Let me know what you think.
